### DISCUSSION TITLE
Inconsistent porosity calculated by OPENPNM and PORESPY

### DISCUSSION THREAD

**Discussion by pengzhulin** [PMEAL/OpenPNM#2823] [Q&A]
I found that porosity calculated by OPENPNM by using a equivalent network from image arrays is different from direct calculation from image array using PORESPY. Here is my script, could you please help?
(1)
im = ps.generators.cylinders(shape=[100,100,100],
                                                 r=int(5),
                                                 porosity=0.7)
im_porosity = ps.metrics.porosity(im)
(2)
'''first get a equivalent network from image arrays'''
snow_output = ps.networks.snow2(im)
pn = op.io.network_from_porespy(snow_output)
'''Then adding some equivalent feature'''
pn.add_model(propname='pore.diameter',
            model=op.models.geometry.pore_size.equivalent_diameter)
pn.add_model(propname='throat.diameter',
            model=op.models.geometry.throat_size.equivalent_diameter)
pn.regenerate_models()

pn.add_model(propname='throat.length',
            model=op.models.geometry.throat_length.cubes_and_cuboids,
            regen_mode='normal')
pn.regenerate_models()

pn.add_model(propname='throat.volume',
            model=op.models.geometry.throat_volume.cuboid,
            regen_mode='normal')
pn.regenerate_models()
''' Then calculated porosity like (https://openpnm.org/examples/applications/effective_diffusivity_and_tortuosity.html)'''
V_p = pn['pore.volume'].sum()
V_t = pn['throat.volume'].sum()
V_bulk = .....
e = (V_p + V_t) / V_bulk

**Comment by jgostick** [comment]
I cannot help, because having a pore network porosity match the image it came from is VERY difficult.  Also, there is already [a page in our examples](https://openpnm.org/examples/applications/porosity.html) on that.

**Comment by pengzhulin** [comment]
Thank you very much for your reply. 

I found that in your source code about Porespy calculating tortuosity_fd like a DNS method. (https://porespy.org/_modules/porespy/simulations/_dns.html#tortuosity_fd). This method doesn't use equivalent but a DNS. Although it cost a lot memory and time to calculate, it seems to be more precise. Will it be a good idea to follow the DNS method by using net = op.network.CubicTemplate(template=im, spacing=1.0)?
